### ISSUE TITLE
[BUG] Baseclass's __cinit__ method does not use subclass's default parameter

### ISSUE THREAD

**Issue by synodriver** [cython/cython#5500]
### Describe the bug

When inherit from a ```cdef class``` with __cinit__ method which hava a default parameter, baseclass's __cinit__ method ignores the defaut parameter from subclass

### Code to reproduce the behaviour:

Considering the following code:
```cython
cdef class Base:
    def __cinit__(self, bint attr = False):
        print(f"in base __cinit__: attr: {attr}")

cdef class Sub(Base):
    def __cinit__(self, bint attr = True):
        print(f"in sub __cinit__: attr: {attr}")
```

### Expected behaviour

Expected:
```
a = Sub()

# in base __cinit__: attr: True
# in sub __cinit__: attr: True 
```

Actual:
```
a = Sub()

# in base __cinit__: attr: False
# in sub __cinit__: attr: True
```

### OS

Windows

### Python version

3.10.8

### Cython version

0.29.35

### Additional context

I know the ```__cinit__``` in baseclass will call first, and using the following code:
```
a= Sub(True)
```
got
> in base __cinit__: attr: True
in sub __cinit__: attr: True

Clearly, the ```__cinit__``` in baseclass will use the parameter passed to the subclass, so why not default parameters?

**Comment by da-woods** [comment]
It isn't completely obvious to me that this is the right thing to do (but equally it isn't obvious that it's wrong). Obviously in regular Python you'd explicitly control this by passing arguments to `super().__init__()`. In Cython we've decided that you can't control and intercept arguments passed to the base class `__cinit__`. Your proposal here would be adding a somewhat obscure exception to the rule where you somehow _can_ influence what gets passed the the base `__cinit__`.

That suggests to me it's the wrong thing to do, because you're going against the general rule on how these methods work.

If we were to do something like this, I think I'd personally rather let the user do an explicit call to `Base.__cinit__` (with the implicit call only generate if the explicit call was missing), and give complete control, rather than adding a slightly obscure back-door mechanism.

One additional thing to consider with your proposal is how you'd like the situation with 3 layers to behave:

```
cdef class L1:
  def __cinit__(self, arg=1):
    print(arg)

cdef class L2(L1):
  def __cinit__(self):  # no arg, is the default from L3 dropped here?
    pass

cdef class L3(L2):
  def __cinit__(self, arg=2):
     print(arg)
```

and there's probably also implementation details about how it should jump between positional and keyword arguments

**Comment by scoder** [comment]
The __cinit__ method is specifically designed as a per-class initialisation step and thus independent of subclasses.

>```cython
>cdef class Base:
>    def __cinit__(self, bint attr = False):
>        print(f"in base __cinit__: attr: {attr}")
>
>cdef class Sub(Base):
>    def __cinit__(self, bint attr = True):
>        print(f"in sub __cinit__: attr: {attr}")
>```

In addition to what David wrote, note that this doesn't even look to be a good design. Breaking the assumptions about a super class in a subclass feels wrong to me and goes against Liskov's substitution principle.

**Comment by synodriver** [comment]
I‘m doing this because sometimes a C level structure should not be initialized in base classes, since they share the same methods, using inherit would be very convenient.
Considering the follwing example:
```
cdef class WrapperA:
    cdef cstructa * a
    def __cinit__(self, bint inita = True):
        if inita:
            self.a = cstructa_new()
        else:
            self.a = cstructb_new()

cdef class WrapperB:
    def __cinit__(self, bint inita = False):
        assert self.a is not NULL
    
```
```__cinit__``` in WrapperA is always invoked first, putting cstructb_new() in WrapperB would causing loss of references to previous structure, adding a way to control the parameters passed to baseclass would be useful.

**Comment by synodriver** [comment]
> One additional thing to consider with your proposal is how you'd like the situation with 3 layers to behave:

I think it should use the parameters from the class that is being called.

> I think I'd personally rather let the user do an explicit call to Base.__cinit__

This is actually what I want. What I'm doing is just a workaround for this

**Comment by da-woods** [comment]
> > somehow can influence what gets passed the the base **cinit**.
> > I think I'd personally rather let the user do an explicit call to Base.**cinit**
> 
> This is actually what I want. What I'm doing is just a workaround for this

If we did do that it'd have to be quite strict - you wouldn't be able to interact with `self` in any way before calling the base `__cinit__`. Because the main point of `__cinit__` is to set up all the invariants that must be in place before interacting with a cdef class.

I personally think this is something we should allow (don't know that everyone agrees though), but it'd have to be implemented pretty strictly and pretty carefully.

**Comment by scoder** [comment]
> the main point of `__cinit__` is to set up all the invariants
> that must be in place before interacting with a cdef class.

Exactly. For everything else, there is already __init__. Just use that if you must respond to subclass needs.

One of the differences between __init__ and __cinit__ is the call order. __cinit__ is called top-down (internally), __init__ is called bottom-up (by the user side code). I don't see why we should change that. Both have different purposes that they seem to serve quite well. And they give different guarantees along the way.

**Comment by synodriver** [comment]
Using ```__init__``` is ok, but users can bypass c structure initialize by using ```sometype.__new__```, which may cause problems.
